### DISCUSSION TITLE
Can't get Serverpod Streaming to work | Keeps on disconnecting

### DISCUSSION THREAD

**Discussion by benwrench41** [serverpod/serverpod#2860] [Q&A] [answered]
Hi, 

I have been stuck on this problem for a few days now.

The idea is, on the client side, a user will update some data, using the company in this example, that update will be sent to the server, then saved to the db and streamed to all users for that company so they have fresh data.

I have a Company Endpoint 

```
class CompanyEndpoint extends Endpoint {
  @override
  bool get requireLogin => true;

  @override
  Future<void> streamOpened(StreamingSession session) async {
    final authenticationInfo = await session.authenticated;
    FullUser? user = await FullUser.db.findFirstRow(
      session,
      where: (p0) => p0.useridId.equals(authenticationInfo?.userId),
    );
    if (user != null) {
      print("User id: ${user.firstName} added to stream");
      print(user.companyId);
      session.messages.addListener(
        user.companyId.toString(),
        (message) {
          print("Message Sent.....");
          sendStreamMessage(session, message);
        },
      );
    }
  }

 Future<void> updateCompany(Session session, Company company) async {
    await Company.db.updateRow(session, company);
    Company? messageToSend = await Company.db.findById(session, company.id!);
    print(messageToSend!.id);

    bool awnt = await session.messages
        .postMessage(messageToSend!.id.toString(), messageToSend);
    print(awnt);
  }}
```
When a user starts up the program they are connected to the endpoints stream, the endpoint finds the user and what company they are with and adds them to the channel for that company. The server prints "User id: Ben added to stream" and "1" so confirmation they are added to the listeners and the channel "companyid", in this case "1". 

Then, from the client, I send the updated Company to the updateCompany function. it updated the db then gets the new company info form the db and posts it to the stream. I get "1" "true" printed by the server, so 1 being the company ID and the true that the message was posted.

then nothing happens. i don't get the "Message Sent....." saying that the message was sent to the stream.

I am handling the stream on the client side with Riverpod here 
```
final companyStreamProvider =
    StateNotifierProvider<CompanyStreamNotifier, AsyncValue<Company?>>((ref) {
  return CompanyStreamNotifier(ref);
});

class CompanyStreamNotifier extends StateNotifier<AsyncValue<Company?>> {
  final Ref ref;
  StreamSubscription<CompanySerial>? _subscription;
  final EndpointCompany streamPoint = client.company;

  CompanyStreamNotifier(this.ref) : super(const AsyncValue.loading());

  Future<void> startListening() async {
    try {
      print("started Listening");
      streamPoint.stream.listen((message) {
        print("message received $message");
        if (message is CompanySerial) {
          ref
              .read(companyProvider.notifier)
              .updateCompanyLocally(message.companyList);
        }
      });
    } catch (e) {
      if (e is StateError) {
        print('Company Stream has already been listened to');
      } else {
        print('Stream Error: $e');
        rethrow;
      }
    }
  }


  @override
  void dispose() {
    print("company stream disposed");
    _subscription?.cancel();
    super.dispose();
  }
}
```
when the app starts I get the message printed "started Listening" and no error. 

Can anyone help? I've read the docs, and used the AI on the Serverpod Docs website and cant find what I'm going wrong. I have tried moving the client listener directly into the widget that's using it, but still nothing.

**Comment by benwrench41** [comment]
So after a bit(lot) of fiddling I've found out that the connection to the server keeps closing. I found out that instead of using the endpoint, I streamed the data to the server and it says not connected, but I have `await client.openStreamingConnection();` in my Main() before the app runs, so it should be connected, so I added this in my stream provider
```
Future<bool> sendUpdatedCompany(Company company) async {
    try {
      await streamPoint.sendStreamMessage(
          CompanySerial(message: "Update", companyList: company));
      print("Sent");
      return true;
    } catch (e) {
      if (e is ServerpodClientException &&
          e.message == 'WebSocket is not connected') {   
        await client.openStreamingConnection();
        await streamPoint.sendStreamMessage(
            CompanySerial(message: "Update", companyList: company));
        print("Sent after reopening connection");
        return true;
      } else {
        rethrow;
      }
    }
  }
```
and it works and sends the message to the server, and the server sends it back, but it prints the message "Sent after reopening connection" and the app doesn't receive it. Each time I click the button it has to connect again, so it probably keeps disconnecting. how do I make it keep the server connected? At the moment I have everything running locally.

**Comment by SandPod** [comment, marked as answer]
Hello @benwrench41!

First, what version of Serverpod are you running? If you are running any version after `2.1.0`, may I suggest using the new Streaming Methods instead? They greatly simplify streaming communication with the server and allow more granular control over your streams to the server.

The documentation is available here: https://docs.serverpod.dev/next/concepts/streams#streaming-methods
We also have a tutorial for building a simple multiplayer drawing app here that should be related to what you are trying to achieve: https://www.youtube.com/watch?v=iCDeAvuMj8I&t=47s

To start debugging you could add a connectivity monitor to the client by calling: `client.connectivityMonitor = <YourConnectivityMonitor>`. We provide a default implementation of the connectivity monitor for flutter named `FlutterConnectivityMonitor` you can use.

We can use that to monitor and detect when the connection is lost (and automatically reconnect). This can help inform us in what cases the connection to the server is lost.

**Reply by benwrench41** [reply]
Thanks for your help, I've updated Serverpod and watched the video. I got rid of the streamOpened override function in my endpoint and replaced it with 
```
  Stream listeningForUpdates(Session session, Company company) async* {
    var updateStream = session.messages.createStream(company.id.toString());
    await for (final update in updateStream) {
      yield update;
    }
  }
```
and changed my stream provider listener to this 
```
Future<void> startListening(Company company) async {
    while (true) {
      try {
        final companyUpdates = client.company.listeningForUpdates(company);
        await for (final update in companyUpdates) {
          if (update is Company) {
            ref.read(companyProvider.notifier).updateCompanyLocally(update);
          }
        }
      } on MethodStreamException catch (e) {
        print(e);
      }

      await Future.delayed(const Duration(seconds: 5));
    }
  }
```

and everything is working perfectly.

**Reply by SandPod** [reply]
Awesome! 🚀

**Reply by chuppasik** [reply]
> Thanks for your help, I've updated Serverpod and watched the video. I got rid of the streamOpened override function in my endpoint and replaced it with
> 
> ```
>   Stream listeningForUpdates(Session session, Company company) async* {
>     var updateStream = session.messages.createStream(company.id.toString());
>     await for (final update in updateStream) {
>       yield update;
>     }
>   }
> ```
> 
> and changed my stream provider listener to this
> 
> ```
> Future<void> startListening(Company company) async {
>     while (true) {
>       try {
>         final companyUpdates = client.company.listeningForUpdates(company);
>         await for (final update in companyUpdates) {
>           if (update is Company) {
>             ref.read(companyProvider.notifier).updateCompanyLocally(update);
>           }
>         }
>       } on MethodStreamException catch (e) {
>         print(e);
>       }
> 
>       await Future.delayed(const Duration(seconds: 5));
>     }
>   }
> ```
> 
> and everything is working perfectly.

Is it necessary to transmit a message from the client to some kind? I don't quite understand how this works.
But for example, if the server updates the data in the database, then I need to get the update on the client.
In this case, the client does not make any changes. What should I do in this case?

**Reply by benwrench41** [reply]
This code I have put here is just to handle the stream. So sending info between the server and clients. it doesn't have any of the other logic

So say, you have 2 clients connected, client A and client B, if client A sends an update to the server, it will get saved in the database, and then the server pushes a message to all the other connected users, so B will also get the update.  In the Endpoint you would have something like this

 ``` 
Future<void> updateCompany(Session session, Company company) async {
    await Company.db.updateRow(session, company);
    Company? messageToSend = await Company.db.findById(session, company.id!);

    await session.messages
        .postMessage(company.id.toString(), messageToSend!, global: false);
  } 
```

 the postMessage will then use the "Stream listeningForUpdates" and send that to the clients. 

The client-side "Future<void> startListening" is a loop that is always on, it will try to connect to the server, if it does, it just sits there and waits. when it gets an update, it does the ```ref.read(companyProvider.notifier).updateCompanyLocally(update);``` command, that will update the riverpod state. 

If it fails to connect it will wait 5 seconds and try to connect again.

They have a lot of excellent information on the serverpod website and other videos. They even have an AI on their website that explain it way better than I can
